README
FriiDump - A program to dump Nintendo Wii and GameCube disc
===============================================================================

FriiDump is a program that lets you dump Nintendo Wii and GameCube disc from
your computer, without using original Nintendo hardware. It basically performs
the same functions as the famous "RawDump" program, but with a big difference,
which should be clear straight from its name: FriiDump is free software, where
"free" is to be intended both as in "free speech" and in "free beer". As such,
FriiDump is distributed with its sources.

This leads to a number of good consequences:
- Having the sources available, it can be easily ported to different operating
  systems and hardware platforms. At the moment it is developed under a
  GNU/Linux system, but it also runs natively on Windows. A MacOS X version can
  be easily created, although I don't have a Mac, so I can't do it myself.
- Also, having the sources and these being well-organized (I know I'm a modest
  guy) allows support for new DVD-ROM drives to be added relatively easily. At
  the moment the same drives as RawDump are supported, but this might improve
  in the future, if anyone takes the effort... See README.technical for
  details.
- The sources might also be used as a reference for several things regarding
  Nintendo Wii/GameCube discs and the hacks used to read them on an ordinary
  drive.

Furthermore, FriiDump also features some functional improvements over RawDump:
- FriiDump can use 4 different methods to read the disc, with different
  performance.
- FriiDump dumps a lot of useful information about the discs it dumps,
  such as whether the disc contains an update or not, which can help avoid
  bricking your Wii ;).
- FriiDump calculates the CRC32, MD4, MD5, SHA-1 and ED2K hashes of dumped
  discs, so you can immediately know if your dump is good or not, by comparing
  the hashes with the well-known ones available on several Internet sites.
- FriiDump comes in the form of a library and a command-line front-end, which
  allows its functions to be easily reused in other programs.

Unfortunately, there is also a main downfall:
- Even the fastest dump method used by FriiDump is not as fast as RawDump (but
  not that much slower, either, see the table below).

Anyway, I'm sure that people who cannot use RawDump (i.e.: GNU/Linux, *BSD and
MacOS X users) will be happy anyway. Besides, you get the sources, so you can
improve them yourself.

Note that FriiDump is only useful to dump *original* Nintendo discs. To dump
backup copies you can use any DVD-dumping program (i.e.: dd under UNIX ;)).

FriiDump came to existance thanks to the work by a lot of people, most of which
are probably not aware of this fact ;). Please see the AUTHORS file for the
credits.


===============================================================================
Supported drives
===============================================================================
At the moment the same drives as RawDump are supported. This is due to various
reasons, explained in the README.technical file, which also contains
information about what is needed to add support for more drives.

Currently supported drives are:
- LG GDR-8161B (untested, but should work)
- LG GDR-8162B (ditto)
- LG GDR-8163B (HL-DT-ST/DVD-ROM GDR8163B)
- LG GDR-8164B (HL-DT-ST/DVD-ROM GDR8164B)

Other drives might work, most likely those based on the Hitachi MN103
microcontroller. If you find any of them, please report so that they can be
added to the compatibility list.

===============================================================================
Docker
===============================================================================

$ docker build -f Docker/Dockerfile -t friidump .
$ docker run --rm --name friidump --privileged -v /dev/sr0:/dev/cdrom friidump -d /dev/cdrom -a

===============================================================================
Installation
===============================================================================
If you are a Windows user, probably you will have downloaded the binaries,
either zipped or together with an installer, so the installation should be
straightforward.

If you downloaded the sources, you will need to compile them. FriiDump uses
CMake, for easy portability, so you will need to get it from cmake.org. On
Windows you will also need a compiler like Visual Studio (the only tested one,
so far) or CygWin/MinGW. On UNIX just do the following, from the directory
where you unpacked the sources into:

$ mkdir BUILD
$ cd BUILD
$ cmake ..
$ make
$ make install

Linux-specific note: You need root privileges to issue certain commands to the
DVD-ROM drive. Hence you have the following possibilities:
- Run FriiDump as root: discouraged.
- Run it through sudo: better but nevertheless discouraged.
- Set the setuid bit on the executable: this is the recommended way to run
  FriiDump under Linux. This way, the code run with superuser privileges will
  be reduced to a minimum, guaranteeing a certain level of security (note that
  security-related bugs might exist anyway!!!). Also note that, even when the
  setuid bit is set, the attempt to open the drive for reading will be done
  after privileges have been dropped, so you will need explicit read access to
  the DVD-ROM drive. Usually having the system administrator add you to the
  "cdrom" group is enough. To set the setuid bit on the executable, run as
  root:

  $ chown root:root /usr/local/bin/friidump
  $ chmod u+s /usr/local/bin/friidump


===============================================================================
Usage
===============================================================================
FriiDump is a command-line program, so you will need to run it from a terminal
or a command-prompt under Windows. The basic usage is as follows:

friidump -d <drive> -a

where <drive> will usually be something like "/dev/hda" on Unix-like systems,
and something like "e:" for Windows users. With this command, the disc will be
dumped to an ISO image file with an automatically-chosen name. Drop -a and use
the -i option if you prefer to specify the filename yourself. If you want to
resume an existing dump, use -s. If you want to dump the disc to a raw format
image file, use -r. Note that you can create a raw and an ISO image at the same
time.

Other options you might want to use are -1 through -4, to set the dump method,
although the default is method 4, which is the fastest one, so most likely you
will not need them.

Finally, use -h for a listing of all available options.


===============================================================================
Performance
===============================================================================
As stated above, FriiDump is not as fast as RawDump. On my PC (Athlon64 3200+),
performance is as follows:

-------------------------------------------------------------------------------
|  Method  |  Dump speed  |  GameCube disc dump time  |  Wii disc dump time   |
-------------------------------------------------------------------------------
|    1     |  Too slow ;) |          Eternity         |   More than eternity  |
|    2     |   ~570 MB/h  |         2.5 hours         |         8 hours       |
|    3     |   ~740 MB/h  |          2 hours          |         6 hours       |
|    4     |  ~1250 MB/h  |         1.2 hours         |        3.5 hours      |
-------------------------------------------------------------------------------


===============================================================================
Support
===============================================================================
I'm releasing this program under the nickname of "Arep". This is because I am
not sure about the legal status of the program, and I do not want to encounter
any consequences. Actually, I'm pretty sure FriiDump goes against the DMCA,
being a program that circumvents copy-protection, but it might be objected that
the format used by Nintendo discs is not a copy-protection method, but just
their own, undocumented, disc format. Although, I think it can be freely used
in Europe and other coutries without laws similar to the DMCA.

For the same reason, I am not putting an e-mail address here (that @no.net you
find in the program is obviously a pun), but support will be provided through
the forums of the Italian ConsoleTribe forum, at http://wii.console-tribe.com.
If you need help, just open a thread in any section there, even in English: I
will *not* reply, but you might stand assured I will read everything you write.
FriiDump users are encouraged to help each other there ;).

Patches are welcome, too: just attach them to your post, and maybe put
something like "[PATCH]" in the topic subject, so that I can easily spot them.

New releases will be announced on that forum, and also on QJ.net, if I find a
good way to notify them.

If you want to donate to the project, do not do it, and donate to one of the
free Wii modchip projects out there, such as OpenWii, WiiFree or YAOSM.



docs/NEWS
Since last official version original method 1 have been renamed to method 0
and it undergone certain changes. Methods 2, 3, 4 have been renamed to 7, 8
and 9 respectively. Method 0 should work with all drives as long as they
are supported by one of memory dump commands, so if drive is unrecognized it
is preferable to keep method at 0, and try all commands. If one of such
combinations turns out to work, you can proceed then testing other methods
with this commad. In case none of commands work, you could try to determine
drive's Read Buffer command's parameters with supplied 'BruteForce3C.exe'.

Generally program's overall bahaviour regarding commandline haven't changed
and you should be able to use same options as with official versions, though
in case you were using unrecognized drive, which would nevertheless work with
Hitachi command, you'll need to set command to 2 now (e.g. --command 2) and
method to 7, 8 or 9.

Performance have increased since official release and should be now about the
same as with 'RawDump'.

Regarding supported drives:

1. Hitachi-LG GDR8161B, GDR8162B, GDR8163B, GDR8164B, GDR8082N
Those drives can read GC/Wii media without swapping. Expected performance is
1600..1900 MB/h for *4B, *3B and 2100..2600 MB/h for *2B, *1B. Custom memory
dump command is used, which returns 2064 bytes of data. It was reproted that
they can not read other (e.g. PC) discs this way though, this needs
confirmation.

2. Lite-On LH-18A1H, DVDRW LH-18A1P, DVDRW LH-20A1H, DVDRW LH-20A1P
Reading performance for PC DVDs can go up to 5000 MB/h, which means program's
core as well as new methods are capable to output data at least at this rate.
Reading performance for GC was about 1600..1700 MB/h so likely this slowdown is
caused by drive logic itself. Though I only had one GC game to test with, so
possibly better results can be achieved depending on media. Best results were
obtained, when using method 5 with parameter 16,27 (--method5=16,27). This
combination isn't set as default because it can cause noticable delays
depending on medium quality and to make methods more general for use with other
devices. Lite-On won't read GC/Wii DVDs at all without swapping. Lite-On
returns 2384 bytes of data (2064 + ECC) by means of vendor specific READ BUFFER
command. Tested with models LH-18A1H, LH-18A1P and LH-20A1H.

3. Plextor
Plextor would return 2064 bytes of already unscrambled data with READ BUFFER
command. It works good with ordinary DVDs but due the lack of streamed reading
support is practically useless for GC/Wii dumping because of very low
performance. Works nevertheless and could be used for some experiments and
testing. Results from PX-760A.

4. Toshiba Samsung SH-D162A, SH-D162B, SH-D162C, SH-D162D
Returns 2384 data bytes per sector like Lite-On does. Appears to support
streamed reading but performance with tested model (SH-D162D) was somewhat low
and unstable even with ordinary DVDs. Looks promising, if only good-working
method could be determined. Latest drives added, definitely need more testing
at this point.


src/friidump.c
/***************************************************************************
 *   Copyright (C) 2007 by Arep                                            *
 *   Support is provided through the forums at                             *
 *   http://wii.console-tribe.com                                          *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#include "misc.h"
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>
#include "disc.h"
#include "dumper.h"
#include "unscrambler.h"

#define USECS_PER_SEC	1000000


/* Name of package */
#define PACKAGE "friidump"

/* Define to the address where bug reports for this package should be sent. */
#define PACKAGE_BUGREPORT "arep@no.net"

/* Define to the full name of this package. */
#define PACKAGE_NAME "FriiDump"

/* Define to the version of this package. */
#define PACKAGE_VERSION "0.5.3.1"


#ifdef WIN32

#include "getopt-win32.h"


#else
#include <sys/time.h>
#include <getopt.h>
#endif


/* Struct for program options */
struct {
	char *device;
	bool autodump;
	bool gui;
	char *raw_in;
	char *raw_out;
	char *iso_out;
	bool resume;
	int dump_method;
	u_int32_t command;
	u_int32_t start_sector;
	u_int32_t sectors_no;
	u_int32_t speed;
	u_int32_t disctype;
	u_int32_t sec_disc;
	u_int32_t sec_mem;
	bool no_hashing;
	bool no_unscrambling;
	bool no_flushing;
	bool stop_unit;
	bool allmethods;
} options;


/* Struct for progress data */
typedef struct {
	struct timeval start_time;
	struct timeval end_time;
	double mb_total;
	double mb_total_real;
	u_int32_t sectors_skipped;
} progstats;


void progress_for_guis (bool start, u_int32_t sectors_done, u_int32_t total_sectors, progstats *stats) {
	int perc;
	double elapsed, mb_done, mb_done_real, mb_hour, seconds_left;
	struct timeval now;
	time_t eta;
	struct tm etatm;
	char buf[50];
	
	if (start) {
		gettimeofday (&(stats -> start_time), NULL);
		stats -> mb_total = (double) total_sectors * 2064 / 1024 / 1024;
		stats -> mb_total_real = (double) (total_sectors - sectors_done) * 2064 / 1024 / 1024;
		stats -> sectors_skipped = sectors_done;
	} else {
		perc = (int) (100.0 * sectors_done / total_sectors);
		gettimeofday (&now, NULL);
		elapsed = difftime (now.tv_sec, (stats -> start_time).tv_sec);
		mb_done = (double) sectors_done * 2064 / 1024 / 1024;
		mb_done_real = (double) (sectors_done - stats -> sectors_skipped) * 2064 / 1024 / 1024;
		mb_hour = mb_done_real / elapsed * 60 * 60;
		seconds_left = stats -> mb_total_real / mb_hour * 60 * 60;
		eta = (time_t) ((stats -> start_time).tv_sec + seconds_left);
		if (localtime_r (&eta, &etatm))
			strftime (buf, 50, "%d/%m/%Y %H:%M:%S", &etatm);
		else
			sprintf (buf, "N/A");

		/* This is the only thing we print to stdout, so that other programs can easily capture and parse our output */
		fprintf (stdout, "%d%%|%u/%u sectors|%.2lf/%.0lf MB|%.0lf/%.0lf seconds|%.2lf MB/h|%s\n",
			 perc, sectors_done, total_sectors, mb_done, stats -> mb_total, elapsed, seconds_left, mb_hour, buf);
		fflush (stdout);
	}

	/* Save return time, in case this will be the last call */
	gettimeofday (&(stats -> end_time), NULL);

	return;
}


void progress (bool start, u_int32_t sectors_done, u_int32_t total_sectors, progstats *stats) {
	int perc, i;
	double elapsed, mb_done, mb_done_real, mb_hour, seconds_left;
	struct timeval now;
	time_t eta;
	struct tm etatm;
	char buf[50];
	
	if (start) {
		gettimeofday (&(stats -> start_time), NULL);
		stats -> mb_total = (double) total_sectors * 2064 / 1024 / 1024;
		stats -> mb_total_real = (double) (total_sectors - sectors_done) * 2064 / 1024 / 1024;
		stats -> sectors_skipped = sectors_done;
	} else {
		perc = (int) (100.0 * sectors_done / total_sectors);
		gettimeofday (&now, NULL);
		elapsed = difftime (now.tv_sec, (stats -> start_time).tv_sec);
		mb_done = (double) sectors_done * 2064 / 1024 / 1024;
		mb_done_real = (double) (sectors_done - stats -> sectors_skipped) * 2064 / 1024 / 1024;
		mb_hour = mb_done_real / elapsed * 60 * 60;
		seconds_left = stats -> mb_total_real / mb_hour * 60 * 60;
		eta = (time_t) ((stats -> start_time).tv_sec + seconds_left);
		if (localtime_r (&eta, &etatm))
			strftime (buf, 50, "%d/%m/%Y %H:%M:%S", &etatm);
		else
			sprintf (buf, "N/A");

		fprintf (stdout, "\r%3d%% ", perc);
		fprintf (stdout, "|");
		for (i = 0; i < 100 / 3; i++) {
			if (i == perc / 3)
				fprintf (stdout, "*");
			else
				fprintf (stdout, "-");
		}
		fprintf (stdout, "| ");
		fprintf (stdout, "%.2lf MB/h, ETA: %s", mb_hour, buf);
		fflush (stdout);
	}

	if (sectors_done == total_sectors)
		printf ("\n");

	/* Save return time, in case this will be the last call */
	gettimeofday (&(stats -> end_time), NULL);

	return;
}



void welcome (void) {
	/* Welcome text */
	fprintf (stderr,
		"FriiDump " PACKAGE_VERSION " - Copyright (C) 2007 Arep\n"
		"This software comes with ABSOLUTELY NO WARRANTY.\n"
		"This is free software, and you are welcome to redistribute it\n"
		"under certain conditions; see COPYING for details.\n"
		"\n"
		"Official support forum: http://wii.console-tribe.com\n"
		"\n"
		"Forum for this UNOFFICIAL VERSION: http://forum.redump.org\n"
		"\n"
		);
	fflush (stderr);

	return;
}


void help (void) {
	/* 80 cols guide:
	 *      |-------------------------------------------------------------------------------|
	 */
	fprintf (stderr, "\n"
		"Available command line options:\n"
		"\n"
		" -h, --help			Show this help\n"
		" -a, --autodump			Dump the disc to an ISO file with an\n"
		"				automatically-generated name, resuming the dump\n"
		"				if possible\n"
		" -g, --gui			Use more verbose output that can be easily\n"
		"				parsed by a GUI frontend\n"
		" -d, --device <device>		Dump disc from device <device>\n"
		" -p, --stop			Instruct device to stop disc rotation\n"
		" -c, --command <nr>		Force memory dump command:\n"
		"				0 - vanilla 2064\n"
		"				1 - vanilla 2384\n"
		"				2 - Hitachi\n"
		"				3 - Lite-On\n"
		"				4 - Renesas\n"
		" -x, --speed <x>		Set streaming speed (1, 24, 32, 64, etc.,\n"
		"				where 1 = 150 KiB/s and so on)\n"
		" -T, --type <nr>		Force disc type:\n"
		"				0 - GameCube\n"
		"				1 - Wii\n"
		"				2 - Wii_DL\n"
		"				3 - DVD\n"
		" -S, --size <sectors>		Force disc size\n"
		" -r, --raw <file>		Output to file <file> in raw format (2064-byte\n"
		"				sectors)\n"
		" -i, --iso <file>		Output to file <file> in ISO format (2048-byte\n"
		"				sectors)\n"
		" -u, --unscramble <file>	Convert (unscramble) raw image contained in\n"
		"				<file> to ISO format\n"
		" -H, --nohash			Do not compute CRC32/MD5/SHA-1 hashes\n"
		"				for generated files\n"
		" -s, --resume			Resume partial dump\n"
		"				-  General  -----------------------------------\n"
		" -0, --method0[=<req>,<exp>]	Use dumping method 0 (Optional argument\n"
		"				specifies how many sectors to request from disc\n"
		"				and read from cache at a time. Values should be\n"
		"				separated with a comma. Default 16,16)\n"
		"				-  Non-Streaming  -----------------------------\n"
		" -1, --method1[=<req>,<exp>]	Use dumping method 1 (Default 16,16)\n"
		" -2, --method2[=<req>,<exp>]	Use dumping method 2 (Default 16,16)\n"
		" -3, --method3[=<req>,<exp>]	Use dumping method 3 (Default 16,16)\n"
		"				-  Streaming  ---------------------------------\n"
		" -4, --method4[=<req>,<exp>]	Use dumping method 4 (Default 27,27)\n"
		" -5, --method5[=<req>,<exp>]	Use dumping method 5 (Default 27,27)\n"
		" -6, --method6[=<req>,<exp>]	Use dumping method 6 (Default 27,27)\n"
		"				-  Hitachi  -----------------------------------\n"
		" -7, --method7			Use dumping method 7 (Read and dump 5 blocks\n"
		"				at a time, using streaming read)\n"
		" -8, --method8			Use dumping method 8 (Read and dump 5 blocks\n"
		"				at a time, using streaming read, using DMA)\n"
		" -9, --method9			Use dumping method 9 (Read and dump 5 blocks\n"
		"				at a time, using streaming read, using DMA and\n"
		"				some speed tricks)\n"
		" -A, --allmethods		Try all known methods and commands until\n"
		"				one works.\n"
#ifdef DEBUG
		" -n, --donottunscramble		Do not try unscrambling to check EDC. Only\n"
		"				useful for testing the raw performance of the\n"
		"				different methods\n"
		" -f, --donottflush		Do not call fflush() after every fwrite()\n"
#endif
	);

	return;
}


bool optparse (int argc, char **argv) {
	bool out;
	char *result = NULL;
	int c;
	int option_index = 0;
	static struct option long_options[] = {
		{"help", 0, 0, 'h'},	//0 - no_argument
		{"autodump", 0, 0, 'a'},
		{"gui", 0, 0, 'g'},
		{"device", 1, 0, 'd'},	//1 - required_argument
		{"raw", 1, 0, 'r'},
		{"iso", 1, 0, 'i'},
		{"unscramble", 1, 0, 'u'},
		{"nohash", 0, 0, 'H'},
		{"resume", 0, 0, 's'},
		{"method0", 2, 0, '0'},	//2 - optional_argument
		{"method1", 2, 0, '1'},
		{"method2", 2, 0, '2'},
		{"method3", 2, 0, '3'},
		{"method4", 2, 0, '4'},
		{"method5", 2, 0, '5'},
		{"method6", 2, 0, '6'},
		{"method7", 0, 0, '7'},
		{"method8", 0, 0, '8'},
		{"method9", 0, 0, '9'},
		{"stop", 0, 0, 'p'},
		{"command", 1, 0, 'c'},
		{"startsector", 1, 0, 't'},
		{"size", 1, 0, 'S'},
		{"speed", 1, 0, 'x'},
		{"type", 1, 0, 'T'},
		{"allmethods", 0, 0, 'A'},
#ifdef DEBUG
		/* We don't want newbies to generate and put into circulation bad dumps, so this options are disabled for releases */
		{"donottunscramble", 0, 0, 'n'},
		{"donottflush", 0, 0, 'f'},
#endif
		{0, 0, 0, 0}
	};

	if (argc == 1) {
		help ();
		exit (1);
	}
	
	/* Init options to default values */
	options.device = NULL;
	options.autodump = false;
	options.gui = false;
	options.raw_in = NULL;
	options.raw_out = NULL;
	options.iso_out = NULL;
	options.no_hashing = false;
	options.resume = false;
	options.dump_method = -1;
	options.command = -1;
	options.start_sector = -1;
	options.sectors_no = -1;
	options.speed = -1;
	options.disctype = -1;
	options.sec_disc = -1;
	options.sec_mem = -1;
	options.no_unscrambling = false;
	options.no_flushing = false;
	options.stop_unit = false;
	options.allmethods = false;

	do {
#ifdef DEBUG
		c = getopt_long (argc, argv, "hpagd:r:i:u:Hs0::1::2::3::4::5::6::789c:t:S:x:T:Anf", long_options, &option_index);
#else
		c = getopt_long (argc, argv, "hpagd:r:i:u:Hs0::1::2::3::4::5::6::789c:t:S:x:T:A", long_options, &option_index);
#endif

		switch (c) {
			case 'h':
				help ();
				exit (1);
				break;
			case 'p':
				options.stop_unit = true;
				break;
			case 'a':
				options.autodump = true;
				options.resume = true;
				break;
			case 'g':
				options.gui = true;
				break;
			case 'd':
				my_strdup (options.device, optarg);
				break;
			case 'r':
				my_strdup (options.raw_out, optarg);
				break;
			case 'i':
				my_strdup (options.iso_out, optarg);
				break;
			case 'u':
				my_strdup (options.raw_in, optarg);
				break;
			case 'H':
				options.no_hashing = true;
				break;
			case 's':
				options.resume = true;
				break;
			case '0':
			case '1':
			case '2':
			case '3':
			case '4':
			case '5':
			case '6':
				options.dump_method = c - '0';
				if (optarg) {
					result = strtok(optarg, ",");
					result = strtok(NULL, ",");
					options.sec_disc = atol(strpbrk(optarg,"1234567890"));
					if (result) options.sec_mem = atol(result);
					else {
						help ();
						exit (1);
					}
				}
				break;
			case '7':
			case '8':
			case '9':
				options.dump_method = c - '0';
				break;
			case 'c':
				options.command = atol (optarg);
				if (options.command > 4) {
					help ();
					exit (1);
				};
				break;
			case 't':
				options.start_sector = atol (optarg);
				break;
			case 'S':
				options.sectors_no = atol (optarg);
				break;
			case 'x':
				options.speed = atol (optarg);
				break;
			case 'T':
				options.disctype = atol (optarg);
				if (options.disctype > 3) {
					help ();
					exit (1);
				};
				unscrambler_set_disctype (options.disctype);
				break;
			case 'A':
				options.allmethods = true;
				options.resume = true;
				break;
#ifdef DEBUG
			case 'n':
				options.no_unscrambling = true;
				break;
			case 'f':
				options.no_flushing = true;
				break;
#endif
			case -1:
				break;
			default:
// 				fprintf (stderr, "?? getopt returned character code 0%o ??\n", c);
				exit (7);
				break;
		}
	} while (c != -1);

	if (optind < argc) {
		/* Command-line arguments remaining. Ignore them, warning the user. */
		fprintf (stderr, "WARNING: Extra parameters ignored\n");
	}

	/* Sanity checks... */
	out = false;
	if (!options.device && !options.raw_in) {
		fprintf (stderr, "No operation specified. Please use the -d or -u options.\n");
	} else if (options.raw_in && options.raw_out) {
		fprintf (stderr,
			"Are you sure you want to convert a raw image to another raw image? ;)\n"
			"Take a look at the -i and -a options!\n"
		);
	} else if (options.autodump && (options.raw_out || options.iso_out)) {
		fprintf (stderr, "The -r and -i options cannot be used together with -a.\n");
	} else {
		/* Specified options seem to make sense */
		out = true;
	}
		
	return (out);
}

int dologic (disc *d, progstats stats) {
	disc_type type_id;
	char *type, *game_id, *region, *maker_id, *maker, *version, *title, tmp[0x03E0 + 4 + 1];
	bool drive_supported;
	int out;
	dumper *dmp;
	u_int32_t current_sector;
	
	

				if (options.stop_unit) { //stop rotation, if requested
					fprintf (stderr, "Issuing STOP command... %s\n", (disc_stop_unit (d, false)) ? "OK" : "Failed");
					exit (1);
				}
				else disc_stop_unit(d, true); //else start rotation

				drive_supported = disc_get_drive_support_status (d);
				fprintf (stderr,
					"\n"
					"Drive information:\n"
					"----------------------------------------------------------------------\n"
					"Drive model........: %s\n"
					"Supported..........: %s\n", disc_get_drive_model_string (d), drive_supported ? "Yes" : "No"
				);

				init_range(d, options.sec_disc, options.sec_mem);

				if (!(disc_set_read_method (d, options.dump_method)))
					exit (2);

				if (options.command!=-1) fprintf (stderr, 
					"Command............: %d (forced)\n", disc_get_command(d));
				else fprintf (stderr, 
					"Command............: %d\n", disc_get_command(d));
				options.dump_method=disc_get_method(d);
				if (disc_get_def_method(d)!=options.dump_method) fprintf (stderr, 
					"Method.............: %d (forced)\n", options.dump_method);
				else fprintf (stderr, 
					"Method.............: %d\n", options.dump_method);
				if ((options.dump_method==0) 
				|| (options.dump_method==1) || (options.dump_method==2) || (options.dump_method==3)
				|| (options.dump_method==4) || (options.dump_method==5) || (options.dump_method==6)
				){
					fprintf (stderr, 
					"Requested sectors..: %d\n", disc_get_sec_disc(d));
					fprintf (stderr, 
					"Expected sectors...: %d\n", disc_get_sec_mem(d));
				}

				fprintf (stderr, "\nPress Ctrl+C at any time to terminate\n");
				fprintf (stderr, "\nRetrieving disc seeds, this might take a while... ");

				//set speed for 1st time
				if (options.speed != -1) disc_set_speed(d, options.speed * 177);
				if (options.speed != -1) disc_set_streaming_speed(d, options.speed * 177);
//				disc_set_speed(d, 0xffff);

				if (!disc_init (d, options.disctype, options.sectors_no)) {
					fprintf (stderr, "Failed\n");
					out = false;
				} else {
					fprintf (stderr, "OK\n");
					disc_get_type (d, &type_id, &type);
					disc_get_gameid (d, &game_id);
					disc_get_region (d, NULL, &region);
					disc_get_maker (d, &maker_id, &maker);
					disc_get_version (d, NULL, &version);
					disc_get_title (d, &title);
					fprintf (stderr, 
						"\n"
						"Disc information:\n"
						"----------------------------------------------------------------------\n");

					if (options.disctype!=-1) fprintf (stderr, 
						"Disc type..........: %s (forced)\n", type);
					else fprintf (stderr, 
						"Disc type..........: %s\n", type);
					if (options.sectors_no!=-1) fprintf (stderr, 
						"Disc size..........: %d (forced)\n", disc_get_sectors_no(d));
					else fprintf (stderr, 
						"Disc size..........: %d\n", disc_get_sectors_no(d));

					if (disc_get_layerbreak(d)>0 && type_id==DISC_TYPE_DVD) fprintf (stderr, 
						"Layer break........: %d\n", disc_get_layerbreak(d));

					if ((type_id==DISC_TYPE_GAMECUBE) || (type_id==DISC_TYPE_WII) || (type_id==DISC_TYPE_WII_DL)) fprintf (stderr, 
						"Game ID............: %s\n"
						"Region.............: %s\n"
						"Maker..............: %s - %s\n"
						"Version............: %s\n"
						"Game title.........: %s\n", game_id, region, maker_id, maker, version, title
					);

					if (type_id == DISC_TYPE_WII || type_id == DISC_TYPE_WII_DL)
						fprintf (stderr, "Contains update....: %s\n" , disc_get_update (d) ? "Yes" : "No");
					fprintf (stderr, "\n");
					
					disc_set_unscrambling (d, !options.no_unscrambling);

					unscrambler_set_disctype (type_id);

					if (options.autodump) {
						snprintf (tmp, sizeof (tmp), "%s.iso", title);
						my_strdup (options.iso_out, tmp);
					}

					//set speed 2nd time after rotation is started and some sectors read
					if (options.speed != -1) disc_set_streaming_speed(d, options.speed * 177);
					if (options.speed != -1) disc_set_speed(d, options.speed * 177);

					/* If at least an output file was specified, proceed dumping, otherwise stop here */
					if (options.raw_out || options.iso_out) {
						if (options.raw_out)
							fprintf (stderr, "Writing to file \"%s\" in raw format\n", options.raw_out);
						if (options.iso_out)
							fprintf (stderr, "Writing to file \"%s\" in ISO format\n", options.iso_out);
						fprintf (stderr, "\n");

						dmp = dumper_new (d);

						dumper_set_hashing (dmp, !options.no_hashing);
						dumper_set_flushing (dmp, !options.no_flushing);

						if (!dumper_set_raw_output_file (dmp, options.raw_out, options.resume)) {
							fprintf (stderr, "Cannot setup raw output file\n");
						} else if (!dumper_set_iso_output_file (dmp, options.iso_out, options.resume)) {
							fprintf (stderr, "Cannot setup ISO output file\n");
						} else if (!dumper_prepare (dmp)) {
							fprintf (stderr, "Cannot prepare dumper");
						} else {
//	 						fprintf (stderr, "Starting dump process from sector %u...\n", dmp -> start_sector);
//		 					opdd.start_sector = options.start_sector;

							if (options.gui)
								dumper_set_progress_callback (dmp, (progress_func) progress_for_guis, &stats);
							else
								dumper_set_progress_callback (dmp, (progress_func) progress, &stats);

							if (dumper_dump (dmp, &current_sector)) {
								fprintf (stderr, "Dump completed successfully!\n");
								if (!options.no_hashing && options.raw_out)
									fprintf (stderr,
										"Raw image hashes:\n"
										"CRC32...: %s\n"
										//"MD4.....: %s\n"
										"MD5.....: %s\n"
										"SHA-1...: %s\n"
										/*"ED2K....: %s\n"*/,
										dumper_get_raw_crc32 (dmp), /*dumper_get_raw_md4 (dmp),*/ dumper_get_raw_md5 (dmp),
										dumper_get_raw_sha1 (dmp)/*, dumper_get_raw_ed2k (dmp)*/
									);
								if (!options.no_hashing && options.iso_out)
									fprintf (stderr,
										"ISO image hashes:\n"
										"CRC32...: %s\n"
										//"MD4.....: %s\n"
										"MD5.....: %s\n"
										"SHA-1...: %s\n"
										/*"ED2K....: %s\n"*/,
										dumper_get_iso_crc32 (dmp), /*dumper_get_iso_md4 (dmp),*/ dumper_get_iso_md5 (dmp),
										dumper_get_iso_sha1 (dmp)/*, dumper_get_iso_ed2k (dmp)*/
									);

								out = true;
								disc_stop_unit (d, 0);
							} else {
								fprintf (stderr, "\nDump failed at sectors: %u..%u\n", current_sector, current_sector+15);
								out = false;
								//disc_stop_unit (d, 0);
							}
						}

						dmp = dumper_destroy (dmp);
					} else {
						fprintf (stderr, "No output file for dumping specified, please take a look at the -i, -r and -a options\n");
					}
				}				
	return out;
}

int main (int argc, char *argv[]) {
	disc *d;
	progstats stats;
	double duration;
	suseconds_t us;
	int out, ret;
	unscrambler *u;
	unscrambler_progress_func pfunc;
	u_int32_t current_sector;

	/* First of all... */
	drop_euid ();
	
	welcome ();

	d = NULL;
	out = false;
	if (optparse (argc, argv)) {
		if (options.device) {
			/* Dump DVD to file */
			fprintf (stderr, "Initializing DVD drive... ");

			if (!(d = disc_new (options.device, options.command))) {
				fprintf (stderr, "Failed\n");
#ifndef WIN32
				fprintf (stderr,
					"Probably you do not have access to the DVD device. Ask the administrator\n"
					"to add you to the proper group, or use 'sudo'.\n"
				);
#endif
			} else {
			fprintf (stderr, "OK\n");
			
				if(options.allmethods)
				{
					fprintf (stderr, "Trying all methods... This will take a LOOOONG time and generate an insanely long console output :p\n");
					for(options.command=0;options.command<=4;options.command++)
					{
						for(options.dump_method=0;options.dump_method<=9;options.dump_method++)
						{
							fprintf (stderr, "Trying with command %d, method %d\n", options.command, options.dump_method);
							out = dologic(d, stats);
							if(out==true)
							{
								break;
							}
						}
						if(out==true)
						{
							fprintf (stderr, "Command %d and method %d combination worked!\n", options.command, options.dump_method);
							break;
						}
					}
				} else {
					out = dologic(d, stats);
				}	
				
				d = disc_destroy (d);
			}
		} else if (options.raw_in) {
			/* Convert raw image to ISO format */
			u = unscrambler_new ();
			
			if (options.gui)
				pfunc = (unscrambler_progress_func) progress_for_guis;
			else
				pfunc = (unscrambler_progress_func) progress;

			if ((out = unscrambler_unscramble_file (u, options.raw_in, options.iso_out, pfunc, &stats, &current_sector)))
				fprintf (stderr, "Unscrambling completed successfully!\n");
			else
				fprintf (stderr, "\nUnscrambling failed at sectors: %u..%u\n", current_sector, current_sector+15);

			u = unscrambler_destroy (u);
		} else {
			MY_ASSERT (0);
		}

		if (out) {
			duration = stats.end_time.tv_sec - stats.start_time.tv_sec;
			if (stats.end_time.tv_usec >= stats.start_time.tv_usec) {
				us = stats.end_time.tv_usec - stats.start_time.tv_usec;
			} else {
				if (duration > 0)
					duration--;
				us = USECS_PER_SEC + stats.end_time.tv_usec - stats.start_time.tv_usec;
			}
			duration += ((double) us / (double) USECS_PER_SEC);
			if (duration < 0)
				duration = 0;
			fprintf (stderr, "Operation took %.2f seconds\n", duration);

			ret = EXIT_SUCCESS;
		} else {
			ret = EXIT_FAILURE;
		}
		
		my_free (options.device);
		my_free (options.iso_out);
		my_free (options.raw_out);
		my_free (options.raw_in);
	}

	return (out);
}
